File: ios/Classes/SoundGeneratorPlugin.swift
import Flutter
import UIKit
import AudioKit

public class SwiftSoundGeneratorPlugin: NSObject, FlutterPlugin {
  var onChangeIsPlaying: BetterEventChannel?;
  var onOneCycleDataHandler: BetterEventChannel?;
  // This is not used yet.
  var sampleRate: Int = 48000;
  var isPlaying: Bool = false;
    
  var oscillator: AKMorphingOscillator = AKMorphingOscillator(waveformArray: [AKTable(.sine),
                                                                                AKTable(.square),
                                                                                AKTable(.triangle),
                                                                                AKTable(.sawtooth),
                                                                              ]);
  var panner: AKPanner?;
  var mixer: AKMixer?;

  public static func register(with registrar: FlutterPluginRegistrar) {
    /*let instance =*/ _ = SwiftSoundGeneratorPlugin(registrar: registrar)
  }

  public init(registrar: FlutterPluginRegistrar) {
    super.init()
    self.panner = AKPanner(self.oscillator, pan: 0.0)
    self.mixer = AKMixer(self.panner!)
    self.mixer!.volume = 1.0
    AKSettings.disableAVAudioSessionCategoryManagement = true
    AKSettings.disableAudioSessionDeactivationOnStop = true
    AKManager.output = self.mixer!
    let methodChannel = FlutterMethodChannel(name: "sound_generator", binaryMessenger: registrar.messenger())
    self.onChangeIsPlaying = BetterEventChannel(name: "io.github.mertguner.sound_generator/onChangeIsPlaying", messenger: registrar.messenger())
    self.onOneCycleDataHandler = BetterEventChannel(name: "io.github.mertguner.sound_generator/onOneCycleDataHandler", messenger: registrar.messenger())
    registrar.addMethodCallDelegate(self, channel: methodChannel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
        case "getPlatformVersion":
          result("iOS " + UIDevice.current.systemVersion)
          break
      case "init":
        //let args = call.arguments as! [String: Any]
        //let sampleRate = args["sampleRate"] as Int
        self.oscillator.frequency = 400
        do {
            try AKManager.start()
            result(true);
        } catch {
            result(FlutterError(
                code: "init_error",
                message: "Unable to start AKManager",
                details: ""))
        }
        break
      case "release":
        result(nil);
        break;
      case "play":
        self.oscillator.start()
        onChangeIsPlaying!.sendEvent(event: true)
        result(nil);
        break;
      case "stop":
        self.oscillator.stop();
        onChangeIsPlaying!.sendEvent(event: false)
        result(nil);
        break;
      case "dB":
        result(self.getDecibel());
        break;
      case "volume":
        result(self.getVolume());
        break;
      case "isPlaying":
        result(self.isPlaying);
        break;
      case "setAutoUpdateOneCycleSample":
        result(nil);
        break;
      case "setFrequency":
        let args = call.arguments as! [String: Any]
        self.oscillator.frequency = args["frequency"] as! Double
        result(nil);
        break;
      case "setWaveform":
        let args = call.arguments as! [String: Any]
        let waveType = args["waveType"] as! String
        
        switch waveType {
            
        case "SINUSOIDAL":
            self.oscillator.index = 0
            
        case "SQUAREWAVE":
            self.oscillator.index = 1
            
        case "TRIANGLE":
            self.oscillator.index = 2
            
        case "SAWTOOTH":
            self.oscillator.index = 3
            
        default:
            self.oscillator.index = 0
        }

        result(nil)
        break;
      case "setBalance":
        let args = call.arguments as! [String: Any]
        self.panner!.pan = args["balance"] as! Double
        break;
      case "setVolume":
        let args = call.arguments as! [String: Any]
        let volume = max(0, min(1, args["volume"] as! Double))
        self.mixer!.volume = volume
        result(nil);
        break;
      case "setDecibel":
        let args = call.arguments as! [String: Any]
        let dB = max(0, min(-20, args["dB"] as! Double))
        let lineerVolume = pow(10.0, dB / 20.0)
        self.mixer!.volume = lineerVolume
        result(nil);
        break;
      case "getSampleRate":
        result(self.sampleRate);
        break;
      case "refreshOneCycleData":
        result(nil);
        break;
      default:
        result(FlutterMethodNotImplemented);
    }
  }
}

File: android/src/main/java/io/github/mertguner/sound_generator/SoundGeneratorPlugin.java
package io.github.mertguner.sound_generator;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import io.github.mertguner.sound_generator.handlers.getOneCycleDataHandler;
import io.github.mertguner.sound_generator.handlers.isPlayingStreamHandler;
import io.github.mertguner.sound_generator.models.WaveTypes;

/** SoundGeneratorPlugin */
public class SoundGeneratorPlugin implements FlutterPlugin, MethodCallHandler {
  private SoundGenerator soundGenerator = new SoundGenerator();
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "sound_generator");
    channel.setMethodCallHandler(this);

    final EventChannel onChangeIsPlaying = new EventChannel(
            flutterPluginBinding.getBinaryMessenger(),
            isPlayingStreamHandler.NATIVE_CHANNEL_EVENT
    );
    onChangeIsPlaying.setStreamHandler(new isPlayingStreamHandler());

    final EventChannel onOneCycleDataHandler = new EventChannel(
            flutterPluginBinding.getBinaryMessenger(),
            getOneCycleDataHandler.NATIVE_CHANNEL_EVENT
    );
    onOneCycleDataHandler.setStreamHandler(new getOneCycleDataHandler());
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      case "getPlatformVersion":
        result.success("Android " + android.os.Build.VERSION.RELEASE);
        break;
      case "init":
        int sampleRate = call.argument("sampleRate");
        result.success(soundGenerator.init(sampleRate));
        break;
      case "release":
        soundGenerator.release();
        break;
      case "play":
        soundGenerator.startPlayback();
        break;
      case "stop":
        soundGenerator.stopPlayback();
        break;
      case "isPlaying":
        result.success(soundGenerator.isPlaying());
        break;
      case "dB":
        result.success(soundGenerator.getDecibel());
        break;
      case "volume":
        result.success(soundGenerator.getVolume());
        break;
      case "setAutoUpdateOneCycleSample":
        boolean autoUpdateOneCycleSample = call.argument("autoUpdateOneCycleSample");
        soundGenerator.setAutoUpdateOneCycleSample(autoUpdateOneCycleSample);
        break;
      case "setFrequency":
        double frequency = call.argument("frequency");
        soundGenerator.setFrequency((float) frequency);
        break;
      case "setWaveform":
        String waveType = call.argument("waveType");
        soundGenerator.setWaveform(WaveTypes.valueOf(waveType));
        break;
      case "setBalance":
        double balance = call.argument("balance");
        soundGenerator.setBalance((float) balance);
        break;
      case "setVolume":
        double volume = call.argument("volume");
        soundGenerator.setVolume((float) volume, true);
        break;
      case "setDecibel":
        double dB = call.argument("dB");
        soundGenerator.setDecibel((float) dB);
        break;
      case "getSampleRate":
        result.success(soundGenerator.getSampleRate());
        break;
      case "refreshOneCycleData":
        soundGenerator.refreshOneCycleData();
        break;
      case "setCleanStart":
        boolean cleanStart = call.argument("cleanStart");
        soundGenerator.setCleanStart(cleanStart);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
